text/0093-metric-instruments.md

```diff
@@ -0,0 +1,104 @@
+# List of new metric instruments
+
+Formalize the new metric instruments proposed in [OTEP 88](https://github.com/open-telemetry/oteps/pull/88).
+
+## Motivation
+
+OTEP 88 introduced a framework for reasoning about new metric
+instruments with various refinements and ended with a [sample
+proposal](https://github.com/open-telemetry/oteps/pull/88#sample-proposal).
+This proposal uses that proposal as a starting point.
+
+## Explanation
+
+The four instrument refinements discussed in OTEP 88 are:
+
+* Sum-only: When computing only a sum is the instrument's primary purpose
+* Non-negative: When negative values are invalid
+* Precomputed-sum: When the application has computed a cumulative sum itself
+* Non-negative-rate: When a negative rate is invalid.
+
+OTEP 88 describes how we are meant to compute rate information from
+instruments having the Non-negative-rate refinement.  Temporal
+aggregation (over time) must be treated as a special case, compared
+with spatial aggregation (over labels).  The logic for computing rates
+depends on whether the Precomputed-sum refinement is present or not,
+which determines whether Delta or Cumulative values are being
+captured.
+
+OTEP 88 proposes that when adding new instruments, we specify
+instruments having a single purpose, each with a distinct set of
+refinements, and each with a carefully selected name for the
+properties of the instrument.
+
+OTEP 88 also also proposes to support language-specific specialization
+as well, to support built-in value types (e.g., timestamps).
+
+## Internal details
+
+The existing specification includes three instruments.  In this
+proposal, the two foundational instruments become abstract, in the
+sense that all synchronous instruments are instances of a Measure
+instrument and all asynchronous instruments are instances of an
+Observer instrument.  New instrument names are given to the
+unrestricted, unrefined versions of rthe foundational instruments:
+
+1. **Distribution** is an an unrefined Measure instrument.  Distribution accepts positive or negative values and uses MinMaxSumCount aggregation by default.
+2. **LastValueObserver** is an unrefined Observer instrument.  LastValueObserver accepts positive or negative values and uses MinMaxSumCount aggregation by default.
```

Comment:
I think this instrument tries to combine actually 2 instruments:
* "DistributionObserver" which accepts a `List<Value>` for every label combination where a MinMaxSumCount makes a lot of sense.
* "LastValueObserver" (GaugeObeserver) which accepts one value per label combination where a "Sum" makes more sense if we talk about merging labels - as an example memory-usage.

Otherwise is confusing to me how the "MinMaxSumCount" aggregation applies to one value reported.

I would suggest that we use "Sum" as default aggregation for this instrument, and add later a DistributionObserver if needed.

Reply:
I do not believe there should be a DistributionObserver. If you want to record a list of values, please use a synchronous instrument.

MinMaxSumCount is a good aggregator for LastValue because it includes the Sum but also characterizes the distribution w/ min/max and average. If you wanted only a Sum aggregation, why not use a DeltaObserver?

MinMaxSumCount is redundant when it includes only a single value, but when pre-aggregating LastValueObserver values you'll get both a distribution and a sum. Exporters that only want to export a sum can do so, the MMSC is cheap enough not to worry about this, I think.

Reply:
This topic is addressed in #96.

Reply:
I've added a note that the names in this document are not finalized.  See #96.